File: Java/Hard/Accepted/remove-max-number-of-edges-to-keep-graph-fully-traversable_3.java
class Solution {
    public int maxNumEdgesToRemove(int n, int[][] edges) {
        
        DSU dsuA = new DSU(n);
        DSU dsuB = new DSU(n);
        int ans = 0;
        for (int[] edge : edges) {
            if (edge[0] == 3) {
                if (!dsuA.union(edge[1] - 1, edge[2] - 1)) {
                    ans++;
                } else {
                    dsuB.union(edge[1] - 1, edge[2] - 1);
                }
            }
        }
        for (int[] edge : edges) {
            if (edge[0] == 1) {
                if (!dsuA.union(edge[1] - 1, edge[2] - 1)) {
                    ans++;
                }
            } else if (edge[0] == 2) {
                if (!dsuB.union(edge[1] - 1, edge[2] - 1)) {
                    ans++;
                }
            }
        }
        if (dsuA.setCount != 1 || dsuB.setCount != 1) {
            return -1;
        }
        return ans;
    }

    class DSU {
        int[] parent;
        int[] rank;
        int setCount;

        public DSU(int n) {
            parent = new int[n];
            rank = new int[n];
            for (int i = 0; i < n; i++) {
                parent[i] = i;
                rank[i] = 1;
            }
            setCount = n;
        }

        public int find(int x) {
            return parent[x] == x ? x : (parent[x] = find(parent[x]));
        }

        public boolean union(int x, int y) {
            int fx = find(x), fy = find(y);
            if (fx == fy) {
                return false;
            }
            if (rank[fx] < rank[fy]) {
                int temp = fx;
                fx = fy;
                fy = temp;
            }
            rank[fx] += rank[fy];
            parent[fy] = fx;
            setCount--;
            return true;
        }
    }
}

File: C++/Hard/CompileError/remove-max-number-of-edges-to-keep-graph-fully-traversable_3.cpp
class Solution {
public:
    int maxNumEdgesToRemove(int n, vector<vector<int>>& edges) {
        
        vector<int> parent(n + 1, 0);
        for (int i = 1; i <= n; ++i) {
            parent[i] = i;
        }
        int result = edges.size();
        for (auto& edge : edges) {
            if (edge[0] == 3) {
                if (find(parent, edge[1]) != find(parent, edge[2])) {
                    unite(parent, edge[1], edge[2]);
                    --result;
                }
            }
        }
        vector<int> parentA(parent);
        for (auto& edge : edges) {
            if (edge[0] == 1) {
                if (find(parentA, edge[1]) != find(parentA, edge[2])) {
                    unite(parentA, edge[1], edge[2]);
                    --result;
                }
            }
        }
        vector<int> parentB(parent);
        for (auto& edge : edges) {
            if (edge[0] == 2) {
                if (find(parentB, edge[1]) != find(parentB, edge[2])) {
                    unite(parentB, edge[1], edge[2]);
                    --result;
                }
            }
        }
        int root = find(parentA, 1);
        for (int i = 2; i <= n; ++i) {
            if (find(parentA, i) != root) {
                return -1;
            }
        }
        root = find(parentB, 1);
        for (int i = 2; i <= n; ++i) {
            if (find(parentB, i) != root) {
                return -1;
            }
        }
        return result;
    }
};